mfai/torch/metrics.py

```diff
@@ -0,0 +1,192 @@
+        self.p = Precision(*args, **kwargs)
+
+    def update(self, *args, **kwargs):
+        self.p.update(*args, **kwargs)
+
+    def compute(self) -> torch.Tensor:
+        return 1 - self.p.compute()
+
+
+class FNR(Metric):
+    """
+    False Negatives Ratio.
+
+    FNR = FN / (TP + FN) = 1 - (TP / (TP + FN)) = 1 - Sensitivity
+    torchmetrics.Sensitivity is not available yet so we implement the calculation.
+    """
+
+    def __init__(self):
+        super().__init__()
+        full_state_update = True  # noqa
+        self.add_state("true_positives", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("false_negatives", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor):
+        assert preds.shape == target.shape
+        preds = torch.where(preds >= 0.5, 1, 0)
+        self.true_positives += torch.sum((preds == 1) & (target == 1))
+        self.false_negatives += torch.sum((preds == 0) & (target == 1))
+
+    def compute(self):
+        return self.false_negatives / (self.true_positives + self.false_negatives)
+
+
+class PR_AUC(Metric):
+    """Area Under the Precsion-Recall Curve."""
+
+    def __init__(self, task: str = "binary"):
+        super().__init__()
+        full_state_update = True  # noqa
+        self.task = task
+
+    def update(self, preds: torch.Tensor, targets: torch.Tensor):
+        pr_curve = PrecisionRecallCurve(task=self.task)
+        self.precision, self.recall, _ = pr_curve(preds, targets)
+
+    def compute(self):
+        return _auc_compute(self.precision, self.recall, reorder=True)
+
+
+class CSINeighborood(Metric):
+    """
+    Compute Criticall Sucess Index (or Threat Score) over a neighborood to avoid the phenomenon
+    of double punishment. So a forecast is considered as a true positive if there is a positive
+    observation in the neighborood (define here by the number of neighbors num_ngb) of a positive
+    prediction.
+
+    For further information on the CSI:
+    https://resources.eumetrain.org/data/4/451/english/msg/ver_categ_forec/uos2/uos2_ko4.htm
+
+    """
+
+    def __init__(
+        self,
+        num_neighbors: int,
+        task: Literal["binary", "multiclass", "multilabel"],
+        num_classes: int = 0,
+        average: Optional[Literal["macro", False]] = "macro",
+    ):
+        super().__init__()
+        self.num_neighbors = num_neighbors
+        self.task = task
+        if task == "binary":
+            self.num_classes = 1
+        elif num_classes == 0 and task != "binary":
+            raise ValueError(
+                "Please define the number of class argument (num_class) when the task is not 'binary'."
+            )
+        else:
+            self.num_classes = num_classes
+        self.average = average
+
+        if torch.cuda.is_available():
+            self.device_csi = torch.device("cuda")
+        else:
+            self.device_csi = torch.device("cpu")
```

Comment:
do we need to define `self.device_csi` ? I think the device is already know by the class in `self.device`, but not 100% sure